### PR TITLE
api: fix some issues in the port redirection

### PR DESCRIPTION
**Pull request by giuseppe**
a few different fixes for the port redirection feature.

### CODE REVIEW

**Comment by giuseppe** [comment]
tests are passing locally

File: api.c [hunk 1]
```diff
@@ -114,7 +114,7 @@ static int api_handle_req_add_hostfwd(Slirp * slirp, int fd, struct api_ctx *ctx
 		free(fwd);
 		goto finish;
 	}
-	if (inet_pton(AF_INET, host_addr_s, &fwd->host_addr) != 1) {
+	if (host_addr_s == NULL || inet_pton(AF_INET, host_addr_s, &fwd->host_addr) != 1) {
```
Comment (AkihiroSuda) [inline api.c]: q: what does `inet_pton` return when `host_addr_s == NULL`?

Comment (giuseppe) [reply api.c]: it segfaults :-)

File: api.c [hunk 2]
```diff
@@ -127,6 +127,10 @@ static int api_handle_req_add_hostfwd(Slirp * slirp, int fd, struct api_ctx *ctx
 		free(fwd);
 		goto finish;
 	}
+
+	if (guest_addr_s == NULL || guest_addr_s[0] == '\0') {
+		guest_addr_s = "10.0.2.100";
```
Comment (AkihiroSuda) [inline api.c]: I'm not huge fan of this, because the guest addr is unknown when slirp4netns was not launched with `-c`

File: api.c [hunk 3]
```diff
@@ -310,5 +314,6 @@ int api_handler(Slirp * slirp, int listenfd, struct api_ctx *ctx)
 	if (rc == 0 && wrc != 0) {
 		rc = wrc;
 	}
+	close(fd);
```
Comment (AkihiroSuda) [inline api.c]: :+1: , thanks

File: api.c (hunk 2 above)
Comment (AkihiroSuda) [reply api.c]: Please add `when guest_addr is empty, guest_addr will be set to the default address that corresponds to --configure` to the man page. Then LGTM

Comment (AkihiroSuda) [reply api.c]: Do you want to define default `host_addr` as well? (0.0.0.0?)

Comment (giuseppe) [reply api.c]: yes good idea, so they behave in a similar way

File: slirp4netns.1.md [hunk 4]
```diff
@@ -143,6 +143,10 @@ $ echo -n $json | nc -U /tmp/slirp4netns.sock
 { "return": {"id": 42}}
 ```
 
+If "host_addr" is not specified, then it defaults to "0.0.0.0".
```
Comment (AkihiroSuda) [inline slirp4netns.1.md]: nit: s/"/**/

**Comment by AkihiroSuda** [comment]
CI failure seems related

File: api.c (hunk 2 above)
Comment (AkihiroSuda) [reply api.c]: The address needs to be now configurable: https://github.com/rootless-containers/slirp4netns/pull/62

Probably `do_slirp` should pass `slirp_config->recommended_vguest` to `api_ctx`

File: api.c (hunk 1 above)
Comment (giuseppe) [reply api.c]: I guess that is expected.  Most C functions segfault if you pass NULL (e.g. strcpy)

**Comment by giuseppe** [comment]
rebased.

The tests flake was not related to the PR.  I've added another patch that seems to fix it on Ubuntu as well

File: tests/test-slirp4netns-api-socket.sh [hunk 5]
```diff
@@ -35,7 +35,7 @@ result=$(echo '{"execute": "add_hostfwd", "arguments":{"proto": "bad"}}' | nc -U
 echo $result | jq .error.desc | grep "bad request: add_hostfwd: bad arguments.proto"
 
 set +e
-result=$(cat /dev/urandom | nc -U $apisocket)
+result=$(head -c 4096 /dev/urandom | nc -U $apisocket)
```
[the remark this answers is not in the document]
Comment (giuseppe) [reply tests/test-slirp4netns-api-socket.sh]: it fails on Ubuntu, I could reproduce it in an Ubuntu VM

[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply tests/test-slirp4netns-api-socket.sh]: Can we leave the test code as-is, and fix slirp4netns code in a separate PR?
Does it fail depending on the length of the input, or the content of the input?

**Comment by AkihiroSuda** [comment]
LGTM except the urandom issue.

Can we have new tests for the issues that are being fixed in this PR? (can be a follow-up PR)

File: tests/test-slirp4netns-api-socket.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply tests/test-slirp4netns-api-socket.sh]: How to reproduce the issue?
Could not reproduce after running test-slirp4netns-api-socket.sh on Ubuntu 18.04 (slirp4netns v0.3.0-alpha.0)

[the remark this answers is not in the document]
Comment (giuseppe) [reply tests/test-slirp4netns-api-socket.sh]: it was introduced by closing the socket after usage.  It seems "nc" from netcat-openbsd doesn't read all the pending data if the socket was closed.

I've changed the implementation of ncat we use for the tests and it seems to work now.

File: tests/test-slirp4netns-api-socket.sh [hunk 6]
```diff
@@ -22,20 +22,20 @@ function cleanup {
...
+result=$(echo '{"execute": "bad"}' | ncat -U $apisocket)
 echo $result | jq .error.desc | grep "bad request: unknown execute"
 
-result=$(echo '{"execute": "add_hostfwd", "arguments":{"proto": "bad"}}' | nc -U $apisocket)
+result=$(echo '{"execute": "add_hostfwd", "arguments":{"proto": "bad"}}' | ncat -U $apisocket)
 echo $result | jq .error.desc | grep "bad request: add_hostfwd: bad arguments.proto"
 
 set +e
-result=$(cat /dev/urandom | nc -U $apisocket)
+result=$(cat /dev/zero | ncat -U $apisocket)
```
Comment (AkihiroSuda) [inline tests/test-slirp4netns-api-socket.sh]: nit: why not urandom?

**Pull request merged**